src/closure.cc

```diff
@@ -18,88 +18,153 @@ using Nan::Persistent;
     #endif
 
-    GIArgument argument;
-    GIArgInfo arg_info;
-    GITypeInfo type_info;
-
-    for (uint i = 1; i < n_param_values; i++) {
-        memcpy(&argument, &param_values[i].data[0], sizeof(GIArgument));
-        g_callable_info_load_arg(closure->info, i - 1, &arg_info);
-        g_arg_info_load_type(&arg_info, &type_info);
-
-        bool mustCopy = true;
-
-        if (signal_query.signal_id) {
-            mustCopy = (signal_query.param_types[i - 1] & G_SIGNAL_TYPE_STATIC_SCOPE) == 0;
-        }
-
-        js_args[i - 1] = GIArgumentToV8(&type_info, &argument, -1, mustCopy);
+    for (uint i = 0; i < nValues; i++) {
+        bool mustCopy = true; // TODO: get information about mustCopy
+        jsArgs[i] = GValueToV8(&values[i], mustCopy);
     }
 
-    Local<Object> self = func;
+    Local<Object> self = fn;
     Local<Value> return_value;
 
     Nan::TryCatch try_catch;
 
-    auto result = Nan::Call(func, self, n_js_args, js_args);
+    auto result = Nan::Call(fn, self, nValues, jsArgs);
 
     if (!try_catch.HasCaught()
             && result.ToLocal(&return_value)) {
-        if (g_return_value) {
-            if (G_VALUE_TYPE(g_return_value) == G_TYPE_INVALID)
+        if (returnValue) {
+            if (G_VALUE_TYPE(returnValue) == G_TYPE_INVALID)
                 WARN ("Marshal: return value has invalid g_type");
-            else if (!V8ToGValue (g_return_value, return_value, true))
+            else if (!V8ToGValue (returnValue, return_value, true))
                 WARN ("Marshal: could not convert return value");
         }
-
-        CallMicrotaskHandlers ();
+        CallMicrotaskHandlers();
     }
     else {
         GNodeJS::QuitLoopStack();
-        Nan::FatalException (try_catch);
+        Nan::FatalException(try_catch);
     }
 
     #ifndef __linux__
-        delete[] js_args;
+        delete[] jsArgs;
     #endif
 }
 
+void Closure::Marshal(GClosure     *base,
+                      GValue       *g_return_value,
+                      uint          n_param_values,
+                      const GValue *param_values,
+                      gpointer      invocation_hint,
+                      gpointer      marshal_data) {
+    auto closure = (Closure *) base;
+    // We don't pass the implicit instance as first argument
+    AsyncCallEnvironment* env = reinterpret_cast<AsyncCallEnvironment *>(Closure::asyncHandle.data);
+    uv_thread_t thread = uv_thread_self();
+    if (uv_thread_equal(&thread, &env->mainThread)) {
+        Closure::Execute(closure->persistent, g_return_value, n_param_values - 1, param_values + 1);
+    } else {
+        CallbackWrapper* cb = new CallbackWrapper();
+        cb->Prepare(&closure->persistent, g_return_value, n_param_values - 1, param_values + 1);
+
+        uv_mutex_lock(&env->mutex);
+        env->queue.push(cb);
+        uv_mutex_unlock(&env->mutex);
+        uv_async_send(&Closure::asyncHandle);
+
+        cb->Wait();
+        delete cb;
+    }
+}
```

Comment:
Yes, you're right. Thanks for the patch. Nice that you also found that `cb` could be stack allocated. I did `delete` it at another place which was terribly wrong and forget to check allocation here.

About the first argument in `param_values`: Of course there's mixing between the logics. It was just a quick idea to get rid of the first argument here because I didn't want to copy it in `Prepare`.

Reply:
Oh yes I see, I hadn't read that part yet when I made this comment. I guess it can be removed before the call dispatch, just make sure to separate the js arguments preparation and the call dispatch.